### DISCUSSION TITLE
How to create a web worker

### DISCUSSION THREAD

**Discussion by tombrouwer** [nuxt/nuxt#16296] [Questions]
Hi all!

I would like to set up a web worker. I'm going to use it as a proxy for my API calls. The user's access token will be loaded into it like an API call. This appears to be safer against XSS.

I have a setup of Nuxt 3 with Typescript. The Vite docs says that I can use `?worker` behind an import. Like:
```
import MyWorker from './worker?worker'

const worker = new MyWorker()
```
(see: [https://vitejs.dev/guide/features.html#import-with-query-suffixes](https://vitejs.dev/guide/features.html#import-with-query-suffixes))

 However, there are two problems:

- The first one, I get a Typescript error: `Cannot find module './path/to/worker?worker' or its corresponding type declarations.`
- There is no worker.js file generated.

Is there another way to compile a web worker and put it as a separate file in the public folder?

**Comment by Qendolin** [comment]
Using the alternative described at https://vitejs.dev/guide/features.html#import-with-constructors worked for me.
I.e.:
```ts
new Worker(new URL('../services/canvas.worker.ts', import.meta.url), {
  type: 'module',
})
```

Note: It did not work without `type: 'module'` and only with a relative path and the `.ts` file extension

**Reply by praveshishere** [reply]
This does not work for me, I'm getting this error 
```js
new URL(url, import.meta.url) is not supported in SSR
```

**Reply by Flanker72** [reply]
URL constructor requires window support.
You may use different approach for client and server like this:
```ts
import url from "node:url";
const myUrl = process.server ? new url.URL(url, import.meta.url) : new window.URL(url, import.meta.url);
```

**Reply by praveshishere** [reply]
Thanks for the help I got it working now.

**Comment by kuizuo** [comment]
```typescript
import MyWorker from './worker.ts?worker'

const worker = new MyWorker()
```

It's working! I have tried.

All the modules used in .ts file are packaged together are packaged as.js file

**Comment by durbanitas** [comment]
Are there any resources about Web Workers and Nuxt 3?

I can only find this thread and a small section about [CDN Edge Workers in the official documentation](https://nuxt.com/docs/guide/concepts/rendering#rendering-on-cdn-edge-workers) that claims Web Workers are available in Nuxt 3.

Is there a good starting point to learn more about Web Workers and how to include them in Nuxt? thanks!

**Comment by neoighodaro** [comment]
Can anyone share something here that helps? So others don’t have to start from scratch. That would be helpful for me and others arriving from Google

**Reply by praveshishere** [reply]
You may use https://nuxt.com/modules/vite-pwa-nuxt, works great.

**Comment by armenr** [comment]
For anyone (including @neoighodaro) who lands here from google...

ASSUMING you are doing things in TypeScript - **and you want to write your worker in TypeScript too** - here's what works: 

1. Create `assets/workers/test.worker.ts`

```ts
// /assets/workers/test.worker.ts
// Listen for messages from the main thread

// Subscribe to messages from main thread
globalThis.onmessage = (event) => {
  try {
    // Parse the JSON string received from the main thread
    const data = JSON.parse(event.data)
    console.log('Received and parsed JSON in worker:', data)

    // Optionally, you can post a message back to the main thread
    globalThis.postMessage('JSON parsed successfully')

    // use setTimeOut to wait 7 seconds and then post a new message to the main thread
    // this should cause the ref in worker.vue to update, and it should reactively change in your browser
    setTimeout(() => {
      globalThis.postMessage('7 seconds have passed')
    }, 7000)
  }
  catch (error) {
    console.error('Error parsing JSON in worker:', error)
  }
}
```

2. create `pages/worker.vue`

```vue
<script setup lang="ts">
import TestWorker from '~/assets/workers/test.worker?worker'

const testWorkerRef = ref('')

onMounted(() => {
  const worker = new TestWorker()

  // Prepare a JSON object
  const message = {
    greeting: 'Hello',
    target: 'World',
  }

  // Convert the object to a JSON string and post it to the worker
  worker.postMessage(JSON.stringify(message))

  // Optionally, handle messages from the worker
  worker.onmessage = (event) => {
    testWorkerRef.value = event.data
    console.log(event.data) // Log the message from the worker
  }
})
</script>

<template>
  <div>
    <h1>
      Hello Nuxt3 with Web Worker and JSON!
    </h1>
    <h1>
      {{ testWorkerRef }}
    </h1>
  </div>
</template>
```

...and that's it. This is a primitive PoC to demonstrate typescript-based web worker in Nuxt3 that works. 🚀😎

**Comment by frenicohansen** [comment]
As stated in [Vite's docs](https://vitejs.dev/guide/assets.html#importing-asset-as-url), we need to include the type from `vite/client` to fix the problem.
> TypeScript, by default, does not recognize static asset imports as valid modules. To fix this, include [vite/client](https://vitejs.dev/guide/features.html#client-types).

And there are two ways to do that according to the docs
> Vite's default types are for its Node.js API. To shim the environment of client side code in a Vite application, add a d.ts declaration file:
> `/// <reference types="vite/client" />`
>
> Alternatively, you can add vite/client to compilerOptions.types inside tsconfig.json:
> 
> ```
> {
>   "compilerOptions": {
>     "types": ["vite/client"]
>   }
> }
> ```
> Source: https://vitejs.dev/guide/features.html#client-types

Nuxt will generate some types definition and already included the type in <workspaceDir>/.nuxt/nuxt.d.ts.
```
// Generated by nuxi

...

/// <reference types="vite/client" />

...

export {}

```

In my case my Nuxt project is part of a monorepo, so it is stored in a subdirectory. And I think typescript failed to get the type from `vite/client`, unless I installed `vite` as a dependency/devDependency in my Nuxt project. It might be because the `vite` package is stored in the monorepo's `node_modules` not in the Nuxt's `node_modules`, if it is not installed as a dependency.

**Reply by al3xnag** [reply]
Same here: monorepo (turborepo) and pnpm. Adding devDependency `vite: *` to nuxt project "fixes" the problem.

**Comment by kytosai** [comment]
I was just wondering, web worker is a pretty important feature when developing web. But why the hell is there absolutely no documentation on nuxt.com about setting up web worker + typescript with nuxt 3?

This is really bad!

**Reply by frenicohansen** [reply]
Daniel has made a Nuxt module to help using web worker https://github.com/danielroe/nuxt-workers

**Reply by kytosai** [reply]
> Daniel has made a Nuxt module to help using web worker https://github.com/danielroe/nuxt-workers

It's still too primitive and lacking features, we can't wait until it's fully developed and stable.

So it's necessary for nuxt.com to have a full documentation guide for developers to use web worker + typescript with nuxt 3 to customize for their purposes!

**Reply by kogratte** [reply]
Moreover, the goal of a webworker is not to be proxified to the main thread, that's what is done with Daniel's package. @kytosai the best documentation you can find about using a worker in nuxt is finally the vite documentation. If you need some example, they're some applications on github using this strategy quiet well (and they'll be more soon I hope)
